File: house-robber/ICE0208.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # 0: nums[0]
        # 1: max(nums[0], nums[1])
        # 2: max(nums[0] + nums[2], nums[1])
        # n : max(dp[n-2] + nums[n], dp[n-1])

        if len(nums) == 1:
            return nums[0]
            
        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i-2] + nums[i] , dp[i-1])
        
        return dp[len(nums) - 1]

File: maximum-subarray/ICE0208.java
class Solution {
    /**
     * 배열을 절반으로 나누고 최대 부분 배열이
     * 왼쪽, 오른쪽, 가운데를 걸치는 경우를 각각 구한다.
     *
     * 시간 복잡도: O(n log n)
     * 공간 복잡도: O(log n)
     */
    public int maxSubArray(int[] nums) {
        return findMaxSum(nums, 0, nums.length - 1);
    }

    private int findMaxSum(int[] nums, int left, int right) {
        if (left == right) {
            return nums[left];
        }

        int mid = left + (right - left) / 2;

        int leftMaxSum = findMaxSum(nums, left, mid);
        int rightMaxSum = findMaxSum(nums, mid + 1, right);
        int crossingMaxSum = findCrossingMaxSum(nums, left, mid, right);

        return Math.max(
            Math.max(leftMaxSum, rightMaxSum),
            crossingMaxSum
        );
    }

    private int findCrossingMaxSum(
        int[] nums,
        int left,
        int mid,
        int right
    ) {
        int sum = 0;
        int leftMaxSum = Integer.MIN_VALUE;

        for (int i = mid; i >= left; i--) {
            sum += nums[i];
            leftMaxSum = Math.max(leftMaxSum, sum);
        }

        sum = 0;
        int rightMaxSum = Integer.MIN_VALUE;

        for (int i = mid + 1; i <= right; i++) {
            sum += nums[i];
            rightMaxSum = Math.max(rightMaxSum, sum);
        }

        return leftMaxSum + rightMaxSum;
    }
}
